Add set_device_map to TensorPipeOptions to support GPU args

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42637 Add set_device_map to TensorPipeOptions to support GPU args**

This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

mrshenli committed:
[WIP] Add set_map_localtion to TensorPipeOptions

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add set_map_localtion to TensorPipeOptions"




[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

[ghstack-poisoned]

mrshenli commented:
@lw since the following messages are expected, should we make it INFO instead of WARNING?

```
[W tensorpipe_agent.cpp:569] RPC agent for worker1 encountered error when reading incoming request from worker0: EOF: end of file (this is expected to happen during shutdown)
[W tensorpipe_agent.cpp:569] RPC agent for worker2 encountered error when reading incoming request from worker1: EOF: end of file (this is expected to happen during shutdown)

```

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli commented:
Stress tests:

After:

```
test/distributed/rpc/test_tensorpipe_agent.py::TensorPipeRpcTestWithSpawn::test_stress_heavy_rpc 
Rank 1 finished testing 20 times in 0.09094452857971191 seconds.
Rank 0 finished testing 20 times in 0.09159255027770996 seconds.
Rank 2 finished testing 20 times in 0.12006711959838867 seconds.
Rank 3 finished testing 20 times in 0.12508106231689453 seconds.
PASSED

test/distributed/rpc/test_tensorpipe_agent.py::TensorPipeRpcTestWithSpawn::test_stress_heavy_rpc_torchscript 
Rank 0 finished testing 20 times in 0.025797605514526367 seconds.
Rank 1 finished testing 20 times in 0.05907726287841797 seconds.
Rank 3 finished testing 20 times in 0.06487584114074707 seconds.
Rank 2 finished testing 20 times in 0.07095670700073242 seconds.
PASSED

test/distributed/rpc/test_tensorpipe_agent.py::TensorPipeRpcTestWithSpawn::test_stress_light_rpc 
Rank 2 finished testing 1000 times in 0.17151093482971191 seconds.
Rank 0 finished testing 1000 times in 0.1786353588104248 seconds.
Rank 3 finished testing 1000 times in 0.18820667266845703 seconds.
Rank 1 finished testing 1000 times in 0.1917405128479004 seconds.
PASSED
```

Before:

```
test/distributed/rpc/test_tensorpipe_agent.py::TensorPipeRpcTestWithSpawn::test_stress_heavy_rpc 
Rank 0 finished testing 20 times in 0.08027839660644531 seconds.
Rank 3 finished testing 20 times in 0.09157371520996094 seconds.
Rank 2 finished testing 20 times in 0.12275886535644531 seconds.
Rank 1 finished testing 20 times in 0.1463322639465332 seconds.
PASSED

test/distributed/rpc/test_tensorpipe_agent.py::TensorPipeRpcTestWithSpawn::test_stress_heavy_rpc_torchscript 
Rank 0 finished testing 20 times in 0.05889773368835449 seconds.
Rank 3 finished testing 20 times in 0.06513261795043945 seconds.
Rank 2 finished testing 20 times in 0.07262659072875977 seconds.
Rank 1 finished testing 20 times in 0.06883549690246582 seconds.
PASSED

test/distributed/rpc/test_tensorpipe_agent.py::TensorPipeRpcTestWithSpawn::test_stress_light_rpc 
Rank 3 finished testing 1000 times in 0.16698956489562988 seconds.
Rank 2 finished testing 1000 times in 0.16751408576965332 seconds.
Rank 1 finished testing 1000 times in 0.18928146362304688 seconds.
Rank 0 finished testing 1000 times in 0.2086775302886963 seconds.
PASSED

```

lw reviewed (commented):
Thanks for doing it! It looks good overall, I left a few comments.

One thing on which I suspect there has been a misunderstanding is how to handle mappings of CPU tensors. In this PR, since we're always copying GPU tensors through the host, it's easy to support GPU-to-CPU transfers or viceversa. But in TensorPipe, at least for the time being, this won't be supported: tensors can only be transferred from CPU to CPU or from a GPU to another GPU. Due to this constraint perhaps, for now, we can remove CPUs from the location mapping entirely, to keep it simpler?

mrshenli commented:
> Due to this constraint perhaps, for now, we can remove CPUs from the location mapping entirely, to keep it simpler?

Yes, let me make that change.

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -93,6 +98,16 @@ def init_rpc(
+        # NB: have to compared .name field here as backend registry can
+        # dynamically modify BackendType object.
+        # Hence, backend == BackendType.TENSORPIPE does not always work.
+        if backend.name == BackendType.TENSORPIPE.name:
```
Yes, good point!

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -15,11 +15,15 @@ def is_available():
-        _enable_jit_rref_pickle,
         _disable_jit_rref_pickle,
+        _enable_jit_rref_pickle,
+        _get_current_rpc_agent,
```
sorry, that's my debug vestige

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -202,6 +203,74 @@ def _all_gather(obj):
+                    )
+
+
+    gathered = _all_gather([torch.cuda.device_count(), map_locations])
```
> An alternative approach could be to have the all-gather collect only the device count and have each worker check only their own location map. Then, in order for all workers to become aware of the validity of other workers' maps, we have a second all-gather pass with the results of the check (thus with just a boolean value).

Yep, I did this in the first version, but realized that we also need to store the reverse mapping. If the quadratic explosion is a concern here, I can do the following:

1. Revert to the two `O(n)` `all_gather` solution.
2. Add an `all_to_all` on top of RPC to collect the reverse map.

Let me know if the above looks better to you.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -202,6 +203,74 @@ def _all_gather(obj):
+        check_one_worker(worker_name, worker_map_locations, all_device_counts)
+
+    # passed all checked, construct reverse mapping for return values
+    reverse_map_locations = {}
```
> I still think that the all-gather is an overkill for this though, because it means that worker #K will also receive the maps between workers #N and #M, even though it doesn't need them. So there is extra data being sent around. What we would need, in principle, is a pairwise swap, so basically an all-to-all collective. But I don't think it makes sense to implement it, let's just roll with all-gather for now.

Let me add an `all_to_all` in the follow up PR. @gqchen would like to know how RPC `all_to_all` compares to c10d `all_to_all` as well. So adding it here would serves two requirements, which seems worth it.

lw replied:
TBH, implementing all-to-all just because of this seems like an overkill: we can stick to the quadratic solution until it becomes a problem and then re-evaluate. If we end up implementing an all-to-all because of other reasons then, at that point, we could consider it here, but we shouldn't make that block the progress on this PR.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -500,6 +500,23 @@ PyObject* rpc_init(PyObject* /* unused */) {
               requests.
+          )")
+      .def_readwrite(
+          "map_locations",
+          &TensorPipeRpcBackendOptions::mapLocations,
+          R"(The device map locations.)")
+      .def(
+          "set_map_location",
```
I like your proposal! :)

@pritamdamania @rohan-varma @osalpekar @jiayisuse @nairbv any suggestion on the naming?

lw replied:
Reading this again, now, maybe we could go further and call it `device_maps`/`set_device_map`?

rohan-varma replied:
I like `set_location_map` and `set_device_map`, the latter also indicates concretely what is being set.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -542,7 +559,11 @@ PyObject* rpc_init(PyObject* /* unused */) {
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "_set_reverse_map_locations",
+          // intentionally not releasing GIL
```
this is very light function, so not releasing GIL to avoid unnecessary context switch. Let me add this to comments

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -92,9 +92,27 @@ struct TensorPipeRpcBackendOptions : public RpcBackendOptions {
 
+  void setMapLocation(
+      const std::string& workerName,
+      const std::unordered_map<c10::DeviceIndex, c10::DeviceIndex>&
```
good point, let me fix.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -202,6 +203,74 @@ def _all_gather(obj):
+                if not all([
+                    len(map_location) == len(key_set),
+                    len(map_location) == len(val_set),  # check 1-to-1 mapping
+                    min(key_set) >= -1,
```
Yep, we can remove that in the first version, but this might be a useful feature in just the driver/master does not have GPU, or if the model cannot fit into its GPU.

lw replied:
Yes, as soon as TensorPipe supports D2H and H2D transfers we should start allowing CPUs in the device map, but until then the only way to do such a transfer is to add an intermediate hop on the CPU (i.e., leader copies across network from its CPU to worker's CPU, and then workers copies locally from CPU to GPU) and this is something that users can already do manually if they so wish.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -31,6 +31,39 @@ const std::string kServerActiveAsyncCalls = "agent.server_active_async_calls";
+    deviceIndices.reserve(tensors.size());
+    const auto& mapLocation = workerIter->second;
+    for (const auto& tensor : tensors) {
+      const auto deviceIter = mapLocation.find(tensor.device().index());
+      if (deviceIter == mapLocation.end()) {
```
Yep, will use `c10::Device` instead of `c10::DeviceIndex`.

mrshenli replied:
Did the check in `options.py` to make sure it has to be CUDA devices.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -202,6 +203,74 @@ def _all_gather(obj):
+                    min(key_set) >= -1,
+                    max(key_set) < device_count,  # check local range
+                    min(val_set) >= -1,
+                    max(val_set) < remote_device_count  # check remote range
```
the current code in the PR allows gaps, and would error out if the callee passes a tensor on an unmapped device. I am open to either. Please let me know which one you prefer.

lw replied:
If we can support it well, I think it's better to allow gaps, as that is the only way to have a device map between workers with different numbers of devices (e.g., imagine a leader having 8 devices and then 8 workers each with only one visible device).

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -606,6 +651,20 @@ Message tensorpipeDeserialize(
+        buffers.deviceIndices.size(),
+        " device indices.");
+    for (size_t i = {}; i < tensors.size(); ++i) {
+      tensors[i] = tensors[i].to(indexToDevice(buffers.deviceIndices[i]));
```
> Calling .to(CPU) for a tensor that is already on CPU is a no-op, right?

Yep, I believe so.

> Do we maybe want to special-case this to avoid this possibly becoming a trap?

Sure, let me add a branch.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -3739,3 +3740,235 @@ def test_tensorpipe_options_throw_on_timedelta_timeout(self):
+    def test_map_locations_gpu(self):
+        options = self.rpc_backend_options
+        dst = worker_name((self.rank + 1) % self.world_size)
+        options.set_map_location(dst, {-1: 0})
```
> Shouldn't this test fail because of the invalid CPU mapping?

Do you mean we should make this fail by disallowing CPU mapping?

lw replied:
I thought that CPUs could only be mapped to CPUs and thus I interpreted this mapping as invalid, but I then realized that we weren't aligned on how CPUs are to be mapped. Since, as I understand it, we're going to disallow CPUs in the device map for now, this will get resolved as part of that.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -3739,3 +3740,235 @@ def test_tensorpipe_options_throw_on_timedelta_timeout(self):
+
+        fn = TensorPipeAgentRpcTest._gpu_add
+        ret = rpc.rpc_sync(dst, fn, args=(torch.zeros(2), torch.ones(2)))
+        self.assertEqual(ret, (torch.zeros(2) + torch.ones(2)).to(0))
```
Yes, I believe so. Let me make the device check more explicit.

rohan-varma commented on test/cpp/rpc/test_tensorpipe_serialization.cpp:
```diff
@@ -127,7 +127,7 @@ TEST(TensorpipeSerialize, NoDeleterTensors) {
   std::tie(sendingTpMessage, tpBuffers) =
       torch::distributed::rpc::tensorpipeSerialize(
-          std::move(sendingRpcMessage));
+          std::move(sendingRpcMessage), {});
```
There are a couple of callsites that pass in empty device indices, what's the semantics in that case (no device mapping is done)? If so, maybe we can put that as a default argument?

mrshenli replied:
Good point. In the current PR, is the map is not specified, users are only allowed to pass CPU tensors (same as the current behavior.)

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -500,6 +500,23 @@ PyObject* rpc_init(PyObject* /* unused */) {
+              function can be called multiple times to incrementally update
+              device placement configurations.
+
+              Arguments:
```
What happens if I call this function when `worker_name` == this worker? Should it work as expected, i.e. if I do an RPC to self and want to map tensors from cuda:0 to cuda:1, will the RPC function run with the tensors on cuda:1? Should we add a test for this and document what's expected/

mrshenli replied:
This should work, but let me add a test for it.

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -500,6 +500,23 @@ PyObject* rpc_init(PyObject* /* unused */) {
+              Arguments:
+                  worker_name (str): Callee name.
+                  map_location (Dict[int, int]): Device placment mappings from
+                  this worker to the callee. This map must be invertible.
```
Is there a test for what happens when the mapping is not invertible?

rohan-varma replied:
This can be ignored, just realized that this is checked at the python level.

rohan-varma commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -31,6 +31,39 @@ const std::string kServerActiveAsyncCalls = "agent.server_active_async_calls";
 const std::string kRpcTimeoutErrorStr =
     "RPC ran for more than set timeout ({} ms) and will now be marked with an error";
 
+std::vector<c10::DeviceIndex> mapTensorDevices(
```
Since this function doesn't actually * do * the mapping (i.e. move the tensors to device) would it be better named something like `getDevicesForTensor`?

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_map_localtion to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -405,7 +445,14 @@ void TensorPipeAgent::pipeWrite(
+      rpcMessage.isRequest() ? opts_.deviceMaps : reverseDeviceMaps_;
+  std::tie(tpMessage, tpBuffers) = tensorpipeSerialize(
+      std::move(rpcMessage),
+      getDevicesForTensors(
```
@lw I noticed that this `TORCH_CHECK` failed in this function, the following line takes a lot longer to finish (probably waiting for timeout). Does it mean we should not crash in `pipeWrite`, and instead should detect the error in the call site?


https://github.com/pytorch/pytorch/blob/0134deda0fb149a8088b15c409617b98c89b34c1/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L808

mrshenli commented:
<img width="856" alt="Screen Shot 2020-08-12 at 6 19 49 PM" src="https://user-images.githubusercontent.com/16999635/90074142-9172a100-dcc8-11ea-97d0-4f0147b7fe65.png">

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -500,7 +479,14 @@ PyObject* rpc_init(PyObject* /* unused */) {
+          &TensorPipeRpcBackendOptions::deviceMaps,
+          R"(The device map locations.)")
+      .def(
+          "set_device_map",
```
is it worth also adding this to the constructor?

lw commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -502,22 +481,12 @@ PyObject* rpc_init(PyObject* /* unused */) {
       .def_readwrite(
-          "map_locations",
-          &TensorPipeRpcBackendOptions::mapLocations,
+          "device_maps",
+          &TensorPipeRpcBackendOptions::deviceMaps,
```
I may be wrong here, but this returns a copy of the maps, right? This means that any change the user does to the returned value, it will not be reflected in the options. In other words:
```
// This has no effect
options.device_maps["worker1"]["cuda:0"] = "cuda:1"
// This works
dev_maps = options.device_maps
dev_maps["worker1"]["cuda:0"] = "cuda:1"
options.device_maps = dev_maps
```
I'm fine with either option, just wanted to make sure this was the behavior you intended.

mrshenli replied:
I tested it, you are right. Which behavior is preferred? I guess we can make it work by returning a `shared_ptr`?

lw commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -173,65 +173,65 @@ def _tensorpipe_construct_rpc_backend_options_handler(
+                    len(device_map) == len(val_set),  # check 1-to-1 mapping
                     min(key_set) >= -1,
                     max(key_set) < device_count,  # check local range
                     min(val_set) >= -1,
```
We said we didn't allow the CPU in the device map, right? If so, these -1 should become 0. And the check below should also go away.

mrshenli replied:
ah, right, forgot to change this part

lw commented on torch/distributed/rpc/options.py:
```diff
@@ -0,0 +1,77 @@
+                )
+            if to in curr_device_maps and k.index in curr_device_maps[to]:
+                curr_v = super().device_maps[to][k.index]
+                raise ValueError(
+                    "`set_device_map` only supports 1-to-1 mapping, "
+                    f"trying to map {k} to {v} and {curr_v}"
+                )
```
Nit: maybe only raise this if `curr_v != v.index`?

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -3873,24 +3859,38 @@ def test_map_locations_multi_gpu(self):
-        dst = worker_name((self.rank + 1) % self.world_size)
-        options.set_map_location(dst, {0: 1})
+
+        options.set_device_map(dst, {0: 1})
```
I must be missing something, but if this is the only device map we define, how can we get back tensors that are allocated on devices 1, 2 and 3?

mrshenli replied:
My bad, these tests are skipped locally as there are only two GPUs. Let me grab a server with more GPUs. hmm, I would expect CI tests cover this, let me check why this is also skipped in CI

mrshenli replied:
Looks like multigpu tests are master-only now.

lw commented on torch/distributed/rpc/options.py:
```diff
@@ -0,0 +1,77 @@
+
+        Arguments:
+            worker_name (str): Callee name.
+            device_map (Dict of int, str, or torch.device): Device placement
```
Does the fact that this accepts a variety of types (int, str, torch.device) cause any problems to TorchScript?

mrshenli replied:
@wanchaol can `Any` type in TorchScript handle this? BTW, how does TorchScript support `torch.device` function?

wanchaol replied:
I think it could but I am not totally sure about it. For `torch.device` we implemented a DeviceObjType in the type system and IValue for c10::Device

mrshenli committed:
Update on "Add set_device_map to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_device_map to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_device_map to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_device_map to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

lw reviewed (approved):
Looks good! Thanks!

mrshenli committed:
Update on "Add set_device_map to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_device_map to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_device_map to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]

mrshenli committed:
Update on "Add set_device_map to TensorPipeOptions to support GPU args"


This commit enables sending non-CPU tensors through RPC using
TensorPipe backend. Users can configure device mappings by calling
`set_map_location` on `TensorPipeRpcBackendOptions`. Internally,
the `rpc_init` API verifies the correctness of device mappings. It
will shutdown RPC if the check failed, or proceed and pass global
mappings to TensorPipeAgent if the check was successful. For serde,
we added a device indices field to TensorPipe read and write buffers,
which should be either empty (all tensors must be on CPU) or match
the tensors in order and number in the RPC message. This commit
does not yet avoid zero-copy, the tensor is always moved to CPU
on the sender and then moved to the specified device on the receiver.

Differential Revision: [D23011572](https://our.internmc.facebook.com/intern/diff/D23011572)

[ghstack-poisoned]
